include/boost/geometry/algorithms/detail/convert_units.hpp

```diff
@@ -0,0 +1,177 @@
+struct convert_coordinate
+{};
+
+template <>
+struct convert_coordinate<degree, radian>
+{
+    template <typename CoordinateType>
+    static inline CoordinateType apply(CoordinateType const& value)
+    {
+        return value * math::d2r<CoordinateType>();
+    }
+};
+
+template <>
+struct convert_coordinate<radian, degree>
+{
+    template <typename CoordinateType>
+    static inline CoordinateType apply(CoordinateType const& value)
+    {
+        return value * math::r2d<CoordinateType>();
+    }
+};
+
+
+template
+<
+    typename UnitsFrom,
+    typename UnitsTo,
+    bool SameUnits = boost::is_same<UnitsFrom, UnitsTo>::value
+>
+struct convert_point
+{
+    template <std::size_t Dimension, typename Point>
+    static inline void convert_units(Point& point)
+    {
+        geometry::set
+            <
+                Dimension
+            >(point,
+              convert_coordinate
+                  <
+                      UnitsFrom, UnitsTo
+                  >::apply(geometry::get<Dimension>(point)));
+    }
+
+    template <typename Point>
+    static inline void apply(Point& point)
+    {
+        convert_units<0>(point);
+        convert_units<1>(point);
+    }
+};
+
+template <typename UnitsFrom, typename UnitsTo>
+struct convert_point<UnitsFrom, UnitsTo, true>
+{
+    template <typename Point>
+    static inline void apply(Point&)
+    {
+    }
+};
+
+
+
+template
+<
+    typename UnitsFrom,
+    typename UnitsTo,
+    bool SameUnits = boost::is_same<UnitsFrom, UnitsTo>::value
+>
+struct convert_box
+{
+    template <typename Box>
+    static inline void apply(Box& box)
+    {
+        detail::indexed_point_view<Box, min_corner> p_min(box);
+        detail::indexed_point_view<Box, max_corner> p_max(box);
+
+        convert_point<UnitsFrom, UnitsTo>::apply(p_min);
+        convert_point<UnitsFrom, UnitsTo>::apply(p_max);
+    }
+};
+
+template <typename UnitsFrom, typename UnitsTo>
+struct convert_box<UnitsFrom, UnitsTo, true>
+{
+    template <typename Box>
+    static inline void apply(Box&)
+    {
+    }
+};
+
+
+}} // namespace detail::units_conversion
+#endif // DOXYGEN_NO_DETAIL
+
+
+#ifndef DOXYGEN_NO_DISPATCH
+namespace dispatch
+{
+
+
+template
+<
+    typename UnitsFrom,
+    typename UnitsTo,
+    typename Geometry,
+    typename Tag = typename tag<Geometry>::type
+>
+struct convert_units
+{};
+
+template <typename UnitsFrom, typename UnitsTo, typename Point>
+struct convert_units<UnitsFrom, UnitsTo, Point, point_tag>
+    : detail::units_conversion::convert_point<UnitsFrom, UnitsTo>
+{};
+
+template <typename UnitsFrom, typename UnitsTo, typename Box>
+struct convert_units<UnitsFrom, UnitsTo, Box, box_tag>
+    : detail::units_conversion::convert_box<UnitsFrom, UnitsTo>
+{};
+
+
+} // namespace dispatch
+#endif // DOXYGEN_NO_DISPATCH
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail
+{
+
+
+template <typename UnitsFrom, typename UnitsTo, typename Geometry>
+inline void convert_units(Geometry& geometry)
```

Comment:
Typically in Geometry similar functions (`convert`, `transform`) has different semantics, i.e. they take two parameters - input and output Geometries. In general the current use case seems to enforce "erroreous" behavior. AFAIU currently this function can be used to convert the coordinates of some Geometry from one unit to another. This means that at some point the new coordinates are calculated and in some cases this geometry contains coordinates in wrong units. Isn't it confusing? Wouldn't it be more clear if a geometry with specific CS and units was used internally and then at the end converted to the output Geometry if needed?